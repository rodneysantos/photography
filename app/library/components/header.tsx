import { clsx } from "clsx";
import { useState } from "react";
import { ChevronDownIcon } from "./chevron-down-icon";
import { Item, Menu } from "./menu";

interface HeaderProps {
	// The function that is called when the active items change.
	onActiveItemsChange: (activeItems: Item[]) => void;
}

export const Header = ({ onActiveItemsChange }: HeaderProps) => {
	const [isMenuOpen, setIsMenuOpen] = useState<boolean>(false);

	// The useCallback hook is used to memoize the function handleMenuVisibility
	// so that it is only re-created when the value of isMenuOpen changes.
	const handleMenuVisibility = () => {
		setIsMenuOpen((prevState) => !prevState);
	};

	return (
		<div className="flex flex-col mt-4 text-center">
			<div className="z-10 mb-4 bg-white">
				<span className="text-4xl tracking-widest font-big-shoulder">
					RODNEY SANTOS
				</span>

				<br />

				<span className="ml-[0.55em] font-dosis text-base tracking-[0.83em]">
					PHOTOGRAPHY
				</span>
			</div>

			<div
				data-testid="menuContainer"
				className={`flex flex-col items-center transition-all will-change-transform ${clsx(
					{
						"-translate-y-14": !isMenuOpen,
					},
				)}`}
			>
				<Menu
					defaultActiveItems={["black and white", "low-key"]}
					onActiveItemsChange={(activeItems) =>
						onActiveItemsChange(activeItems)
					}
					className={`transition-opacity will-change-auto ${clsx({
						"opacity-0": !isMenuOpen,
					})}`}
				/>

				<ChevronDownIcon
					className={`w-5 mt-4 transition-opacity cursor-pointer will-change-auto ${clsx(
						{
							"opacity-20": !isMenuOpen,
						},
					)}`}
					onClick={handleMenuVisibility}
				/>
			</div>
		</div>
	);
};
