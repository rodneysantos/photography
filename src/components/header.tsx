import { createSignal } from "solid-js";
import { ChevronDownIcon } from "./chevron-down-icon";
import { Item, Menu } from "./menu";

interface HeaderProps {
	// The function that is called when the active items change.
	onActiveItemsChange: (activeItems: Item[]) => void;
}

export const Header = ({ onActiveItemsChange }: HeaderProps) => {
	const [isMenuOpen, setIsMenuOpen] = createSignal(false);

	// The useCallback hook is used to memoize the function handleMenuVisibility
	// so that it is only re-created when the value of isMenuOpen changes.
	const handleMenuVisibility = () => {
		setIsMenuOpen((prevState) => !prevState);
	};

	return (
		<div class="flex flex-col text-center 4xl:mt-14 5xl:mt-16 5xl:mb-4">
			<div class="z-10 mb-4 bg-white">
				<div class="tracking-widest 3xl:text-3xl 4xl:text-4xl 5xl:text-5xl font-big-shoulder">
					RODNEY SANTOS
				</div>

				<div class="text-base font-dosis 5xl:text-2xl 4xl:text-xl 4xl:ml-[0.37rem] 4xl:tracking-[0.69rem] 5xl:ml-5xl 5xl:tracking-5xl">
					PHOTOGRAPHY
				</div>
			</div>

			<div
				data-testid="menuContainer"
				class="flex flex-col items-center transition-all will-change-transform"
				classList={{ "-translate-y-14": !isMenuOpen() }}
			>
				<Menu
					defaultActiveItems={["black and white", "low-key"]}
					onActiveItemsChange={onActiveItemsChange}
					show={isMenuOpen()}
				/>

				<ChevronDownIcon onClick={handleMenuVisibility} show={!isMenuOpen()} />
			</div>
		</div>
	);
};
