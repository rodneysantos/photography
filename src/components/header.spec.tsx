import { cleanup, fireEvent, render } from "@solidjs/testing-library";
import { afterEach, beforeEach, describe, expect, it, vi } from "vitest";
import { Header } from "./header";

describe("header", () => {
	const onActiveItemsChange = vi.fn();

	beforeEach(() => {
		onActiveItemsChange.mockReset();
	});

	afterEach(cleanup);

	it("renders correctly", () => {
		// arrange
		const { asFragment } = render(() => (
			<Header onActiveItemsChange={onActiveItemsChange} />
		));

		// assert
		expect(asFragment()).toMatchSnapshot();
	});

	it("animates the menu container on click", () => {
		// arrange
		const { getByTestId } = render(() => (
			<Header onActiveItemsChange={onActiveItemsChange} />
		));

		// act
		fireEvent.click(getByTestId("chevronDownIcon"));

		// assert
		expect(getByTestId("menuContainer").className).not.toContain(
			"-translate-y-14",
		);
		expect(getByTestId("chevronDownIcon").getAttribute("class")).toContain(
			"opacity-20",
		);
	});

	it("emits the active items on change", () => {
		// arrange
		const { getByTestId } = render(() => (
			<Header onActiveItemsChange={onActiveItemsChange} />
		));

		// act
		const blackAndWhiteItem = getByTestId("bnw");
		blackAndWhiteItem.click();

		// assert
		expect(onActiveItemsChange).toHaveBeenCalledTimes(1);
	});
});
