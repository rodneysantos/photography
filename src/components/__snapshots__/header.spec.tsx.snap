// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`header > renders correctly 1`] = `"<div class=\\"flex flex-col text-center 4xl:mt-14 5xl:mt-16 5xl:mb-4\\"><div class=\\"z-10 mb-4 bg-white\\"><div class=\\"tracking-widest 3xl:text-3xl 4xl:text-4xl 5xl:text-5xl font-big-shoulder\\">RODNEY SANTOS</div><div class=\\"text-base font-dosis 5xl:text-2xl 4xl:text-xl 4xl:ml-[0.37rem] 4xl:tracking-[0.69rem] 5xl:ml-5xl 5xl:tracking-5xl\\">PHOTOGRAPHY</div></div><div data-testid=\\"menuContainer\\" class=\\"flex flex-col items-center transition-all will-change-transform -translate-y-14\\"><div data-testid=\\"menu\\" class=\\"transition-opacity will-change-auto opacity-0\\"><div class=\\"flex flex-row justify-center\\"><div class=\\"px-2 py-1 mx-2 text-center border-2 cursor-pointer border-neutral-900 bg-neutral-900 text-neutral-100\\" data-testid=\\"black and white\\">black and white</div><div class=\\"px-2 py-1 mx-2 text-center border-2 cursor-pointer border-neutral-900\\" data-testid=\\"color\\">color</div><div class=\\"px-2 py-1 mx-2 text-center border-2 cursor-pointer border-neutral-900 bg-neutral-900 text-neutral-100\\" data-testid=\\"low-key\\">low-key</div></div></div><svg data-testid=\\"chevronDownIcon\\" xmlns=\\"http://www.w3.org/2000/svg\\" viewBox=\\"0 0 512 512\\" class=\\"w-5 mt-4 transition-opacity cursor-pointer will-change-auto\\"><path d=\\"M256 0a256 256 0 1 0 0 512A256 256 0 1 0 256 0zM135 241c-9.4-9.4-9.4-24.6 0-33.9s24.6-9.4 33.9 0l87 87 87-87c9.4-9.4 24.6-9.4 33.9 0s9.4 24.6 0 33.9L273 345c-9.4 9.4-24.6 9.4-33.9 0L135 241z\\"></path></svg></div></div>"`;
